### DISCUSSION TITLE
Manuelle Programmausführung - "Aktiv"-Status wird (nicht) berücksichtigt

### DISCUSSION THREAD

**Discussion by jp112sdl** [OpenCCU/OpenCCU#1337] [🇩🇪 Fragen&Antworten]
Der standardmäßig mitgebrachte Button zum manuellen Ausführen eines Programms löst die Abarbeitung des "DANN"-Blocks immer aus - <ins>unabhängig</ins> davon, ob das Programm "aktiv" oder "inaktiv" ist.
Das ist auch gut so, denn so kann man trotz abgestellter Automatik eine Ausführung forcieren.

Der in #1285 neu eingebaute Button zum manuellen Ausführen eines Programms mit Bedingungsprüfung führt das Programm nur aus, wenn es "aktiv" ist.

Soll letzteres so bleiben?
Wenn nein - welche Möglichkeiten gibt es?
Die Ausführung mit Bed.-Prüfung wird mittels `.State(1)` ausgelöst.
`State` scheint hier also den `Active`-State zu prüfen.

Nun könnte man zunächst prüfen, ob das Programm aktiv ist.
Wenn nicht - dann setze es aktiv, führe es aus, setze es inaktiv.
Jedoch scheint schon der Wechsel von aktiv->inaktiv das Programm zu triggern.

Oder gibt es die Möglichkeit, noch einen Parameter an `.State(1,xxx)` anzuhängen, mit dem man das Ausführen eines inaktiven Programms erzwingen kann?

**Comment by jens-maus** [comment]
Programme werden innerhalb der ReGaHss im Grunde wie Datenpunkte von Geräten gehandhabt und als solches wird da natürlich auch das active flag geprüft um ein setzen des `State()` zu verhindern wenn er im Grunde unnötig ist. Wenn es also gewollt ist das man ein Program jederzeit ausführen kann egal ob es als "aktiv" oder "nicht aktiv" markiert ist müsste man wie von dir beschrieben vorgehen und das Program vorher auf aktiv setzen, ja. Wenn allerdings bereits das aktiv->inaktiv setzen das Programm triggert ist das natürlich potentiell kritisch, ja. Eine wirkliche Lösung habe ich aber gerade nicht parat und nein, es gibt auch keine zusätzliche `State()` option um die ausführung eines Programmes zu erzwingen trotz inactive status.

**Reply by jp112sdl** [reply]
Darüber gestolpert bin ich eigentlich nur, weil ich in der Programm-Editier-Ansicht im Footer auch noch einen "Ausführen" Button einbauen wollte. 
So könnte man das soeben erstellte/bearbeitete Programm direkt testen, ohne es speichern und in "Status und Bedienung" wechseln zu müssen. _(Das spart eine Menge Klickerei, wenn man nicht mit mehreren Tabs arbeiten möchte (was die WebUI ja auch nicht so gern hat).)_

Jedoch wird ja ein editiertes Programm als Kopie mit Status "inaktiv" angelegt.
Und das lässt sich mit `State(1)` nicht ausführen.

Eine Lösung fällt mir dazu dann aber auch nicht ein.

Aber vielleicht sollte der "Ausführen - nur 1. DANN" Button nicht sichtbar/deaktiviert sein, wenn das Programm nicht "aktiv" ist. Es suggeriert sonst eine tatsächliche Ausführung

**Comment by Baxxy13** [comment]
Hmm,
sicher das das Programm bei Aktivierung ausgeführt wird? Der Zeitstempel wird hier zwar aktualisiert aber obwohl Bedingungen erfüllt sind geht es weder ins DANN noch ins SONST. Grob getestet mit "nur prüfen", "bei Änderung" und "bei Aktualisierung".
Müsste aber noch verifiziert werden. 
Ist dem so, könnte man mit dem neuen Butten...

- das Programm aktivieren wenn inaktiv und dann ausführen mit Prüfung 

Ausblenden wäre zwar auch ne Option, gefiele mir aber weniger.

**Reply by jp112sdl** [reply]
> sicher das das Programm bei Aktivierung ausgeführt wird?

<img src="http://g.recordit.co/ONR7iAFrMO.gif" />

**Reply by Baxxy13** [reply]
Mein Kaffee war heute morgen wohl schon zu kalt und ich habe mich selbst überlistet. ;-)
Wenn ich mit...
```
object dp = dom.GetObject("$src$");
if (dp) {...
```
... logge kann natürlich nix kommen da es keinen reellen Trigger gibt. 

Das Gute ist zumindest das beim Aktivieren das Programm mit Bedingungsprüfung startet.

Somit könnte man (wenn es möglich wäre) beim Druck auf den neuen Button prüfen welchen Status das Programm hat (aktiv / inaktiv) und entsprechend entweder `Active(true)` oder `.State(1)` ausführen. Bei ersterem liegt es natürlich dann beim User danach das Programm (wenn nötig) wieder zu deaktivieren.

**Reply by blackbasket** [reply]
Mein Kaffee ist immer kalt, bis ich dazu komme, ihn zu trinken ;)

Ich denke, dass es "gefährlich" ist, das Aktiv-Flag über den Button zu setzen. Weil eben das bewusste Deaktivieren danach schnell vergessen ist.

**Reply by jens-maus** [reply]
So sehe ich das auch. Sowas sollte man nicht tun. Somit wäre das wohl IMHO nur noch eine möglicher Feature Request für eine zukünftige Version ReGaHss wo ich dann ggf. etwas einbauen kann das man auch deaktivierte programme ablaufen lassen kann.

**Reply by jp112sdl** [reply]
Ich hätte sonst noch vorgeschlagen:

- Vor der Ausführung wird der "aktiv"-Status geprüft und
  - ein aktives Programm wird (wie derzeit implementiert) mit `objPrg.State(1);` ausgeführt
  - ein inaktives Programm wird mit `objPrg.Active(true); objPrg.Active(false);` ausgeführt. _(quasi ein kurzes Toggeln des aktiv-Flags)_

**Reply by Baxxy13** [reply]
Kurzes toggeln als Workaround klingt erstmal gut. 
Langfristig wäre aber eine Implementierung einer neuen Methode die sauberste Lösung. Dann könnte man ja auch ein inaktives Programm bei Bedarf per Script mit Bedingungsprüfung starten. Mir fällt zwar spontan kein Szenario dafür ein, aber haben ist ja bekanntlich besser als brauchen.

**Reply by jp112sdl** [reply]
> Langfristig wäre aber eine Implementierung einer neuen Methode die sauberste Lösung.

Ja, sehe ich auch so.

@jens-maus : Ist es sinnvoll, zunächst die `.Active(true/false)`-"Toggle"-Lösung einzubauen oder hast du die _"...zukünftige Version ReGaHss wo ich dann ggf. etwas einbauen kann das man auch deaktivierte programme ablaufen lassen kann..."_ zum nächsten RM Release bereits eingeplant?

**Reply by jp112sdl** [reply]
P.S.: Ich habe die Änderung (ist ja doch nur minimal) als Review an den [Patch](https://github.com/jens-maus/RaspberryMatic/pull/1291/files) angehängt.

Ich wollt jetzt für den Einzeiler nicht noch mal einen kompletten PR erstellen.
Bei "Gefallen" kann er gern von Jens direkt ins File übernommen werden.

**Reply by Baxxy13** [reply]
> P.S.: Ich habe die Änderung (ist ja doch nur minimal) als Review an den [Patch](https://github.com/jens-maus/RaspberryMatic/pull/1291/files) angehängt.

Kannst du mal bitte ein .patch-File erstellen damit ich das testen kann? 
Irgendwie bekomme ich das nicht hin. 
Wenn ich den RAW-Code als Patchfile speichere kommt beim patchen mehrfach... `Hunk 1 FAILED`

Ich und Linux --- grrrr

**Reply by jp112sdl** [reply]
@Baxxy13 Hast du den alten Patch vorher revoked ?

Vielleicht ist es hier jetzt mal einfacher, wenn du in der `/www/webui/webui.js` die Zeile
` pb += (thenOnly === true) ? "  o.ProgramExecute();" : "  o.State(1);";`
von Hand änderst in
`pb += (thenOnly === true) ? " o.ProgramExecute();" : " if (o.Active() == true) { o.State(1); } else { o.Active(true); o.Active(false); }";`

**Reply by Baxxy13** [reply]
> Hast du den alten Patch vorher revoked ?

Ach iwo. ;-) 
Einfach versucht drüberzubügeln. Ich setzte mal die .ova neu auf und teste dann händisch.

**Reply by jens-maus** [reply]
Hab mich ja noch nicht weiter dazu geäußert, aber den vorgeschlagenen Workaround via Inactive…Active sehe ich kritisch. Auch wenn das natürlich prinzipiell geht halte ich das nichz für eine gute idee das generell so umzusetzen.

**Reply by Baxxy13** [reply]
> Workaround via Inactive…Active sehe ich kritisch

Na dann... her mit einer sicheren Methode. ;-)



> Review

Nun erfolgreich implementiert und getestet. Man merkt ja nicht mal das das Programm kurz aktiv geschalten wird.
Von daher für mich als Workaround in Ordnung.

**Reply by jp112sdl** [reply]
Mir geht es funktional hauptsächlich darum, das Programm in der Editor-Ansicht (wo man es als inaktive Kopie bearbeitet), bereits ausführen zu können.

<img width="1064" alt="Bildschirmfoto 2021-07-11 um 13 11 38" src="https://user-images.githubusercontent.com/18190139/125192777-cd501680-e249-11eb-8c33-c6d7367e95c4.png">

Der Button ist auch nur ein Einzeiler in der `/www/rega/pages/tabs/admin/views/programs.htm`, der die temporäre Programm-ID der `ExecuteProgram`-Methode in der `webui.js` übergibt.

Falls @Baxxy13 das auch testen möchte:
`    s += "<td style='text-align:center; vertical-align:middle;'><div class='FooterButton' onclick='ExecuteProgram(<% Write(oPrg.ID()); %>);' >${actionStatusControlLblStart}</div></td>";` 
einfügen z.B. nach [Zeile 947](https://github.com/eq-3/occu/blob/f88d81a54e004fb895efbe7ec829c4d1ef28dd47/WebUI/www/rega/pages/tabs/admin/views/programs.htm#L947)

**Reply by Baxxy13** [reply]
> das auch testen möchte:

Gesagt, getan. Den Gedanken dahinter finde ich sehr gut. 
Button ist drin... aber funktioniert das bei Dir?
![JP_Test_Button](https://user-images.githubusercontent.com/54898249/125194339-23748800-e251-11eb-8729-f59f7ee35cf3.JPG)
Da tut sich nix im Protokoll, egal ob die Logik-SV wahr oder unwahr ist.

**Reply by jens-maus** [reply]
> Nun erfolgreich implementiert und getestet. Man merkt ja nicht mal das das Programm kurz aktiv geschalten wird.

Externe Apps die auf den status von objekten/programmen, etc. horchen bekommen das aber mit! Z.b. ioBroker

> Von daher für mich als Workaround in Ordnung.

für mich aber weiterhin nicht und ich präferiere weiterhin das direkt in der rega einzubauen

**Reply by jp112sdl** [reply]
> aber funktioniert das bei Dir?

@Baxxy13 Prüfe ich später noch mal. Ich war der Meinung, dass es ging.

>  ich präferiere weiterhin das direkt in der rega einzubauen

@jens-maus das wäre auch meine Lieblingslösung.
Jedoch kann ich den Aufwand nicht abschätzen.
